_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,386 @@
+  const convectionTypeDefs = readFileSync("src/data/convection.graphql", "utf8")
+
+  // Setup the default Schema
+  const schema = await makeRemoteExecutableSchema({
+    schema: convectionTypeDefs,
+    link: convectionLink
+  })
+
+  // Remap the names of certain types from Convection to fit in the larger
+  // metaphysics ecosystem.
+  const remap = {
+    Submission: "ConsignmentSubmission",
+    Category: "ConsignmentSubmissionCategoryAggregation",
+    Asset: "ConsignmentSubmissionCategoryAsset",
+    State: "ConsignmentSubmissionStateAggregation",
+    SubmissionConnection: "ConsignmentSubmissionConnection"
+  }
+
+  // Return the new modified schema
+  return transformSchema(schema, [
+    new RenameTypes(name => {
+      const newName = remap[name] || name
+      return newName
+    })
+  ])
+}
+```
+
+Another example is to outright remove almost everything in the schema, and to only allow Types and fields which we
+know to be useful.
+
+```js
+export const executableGravitySchema = async () => {
+  const gravityTypeDefs = readFileSync("src/data/gravity.graphql", "utf8")
+
+  const gravityLink = createGravityLink()
+  const schema = await makeRemoteExecutableSchema({
+    schema: gravityTypeDefs,
+    link: gravityLink,
+  })
+
+  // Types which come from Gravity which MP already has copies of.
+  // In the future, these could get merged into the MP types.
+  const blacklistedTypes = ["Artist", "Artwork"]
+
+  // Gravity's GraphQL contains a bunch of objects and root fields that will conflict
+  // with what we have in MP already, this lets us bring them in one by one
+  const whitelistedRootFields = ["Query", "recordArtworkView"]
+
+  // Return the new modified schema
+  return transformSchema(schema, [
+    new FilterRootFields((_type, name) => {
+      return !whitelistedRootFields.includes(name)
+    }),
+    new FilterTypes(type => {
+      return !blacklistedTypes.includes(type.name)
+    }),
+    // snip
+  ])
+})
+```
+
+We can write tests for this by running [a query which returns all of the types][type-q] in a schema, and validating
+what exists:
+
+```js
+import { executableGravitySchema } from "../schema"
+import { getTypesFromSchema } from "lib/stitching/lib/getTypesFromSchema"
+
+it("Does not include blacklisted types", async () => {
+  const gravitySchema = await executableGravitySchema()
+  const gravityTypes = await getTypesFromSchema(gravitySchema)
+
+  expect(gravityTypes).not.toContain("Artist")
+  expect(gravityTypes).not.toContain("Artwork")
+})
+```
+
+This one is interesting, we _don't_ want the version of `Artist` and `Artwork` from Gravity's GraphQL
+implementation - because the hand-rolled `Artwork` and `Artist` types which lives in the source code of Metaphysics
+right now is a combination of many sources, and front-end-client specific code. If we allowed the `Artist` or
+`Artwork` to overwrite the existing implementations it would be a massively breaking change.
```

Comment:
This problem really highlights the problems of smashing schemas together without care; a connecting sentence after this paragraph could help readers understand the distinction between merging and stitching later.

Reply:
great point, making that  really specific definitely added